resharper/resharper-unity/src/Rider/Shaders/ShaderContextCache.cs

```diff
@@ -0,0 +1,80 @@
+using JetBrains.Diagnostics;
+using JetBrains.DocumentManagers;
+using JetBrains.DocumentModel;
+using JetBrains.Lifetimes;
+using JetBrains.ProjectModel;
+using JetBrains.ReSharper.Feature.Services.Daemon;
+using JetBrains.ReSharper.Plugins.Unity.HlslSupport.Integration.Injections;
+using JetBrains.ReSharper.Psi;
+using JetBrains.ReSharper.Psi.Caches;
+using JetBrains.ReSharper.Psi.Cpp.Caches;
+using JetBrains.ReSharper.Resources.Shell;
+using JetBrains.Util;
+using JetBrains.Util.Caches;
+
+namespace JetBrains.ReSharper.Plugins.Unity.Rider.Shaders
+{
+    [SolutionComponent]
+    public class ShaderContextCache : IPreferredRootFileProvider
+    {
+        private readonly ISolution mySolution;
+        private readonly InjectedHlslFileLocationTracker myLocationTracker;
+        private readonly DocumentManager myManager;
+        private readonly ILogger myLogger;
+        private readonly DirectMappedCache<IPsiSourceFile, IRangeMarker> myShaderContext = new DirectMappedCache<IPsiSourceFile, IRangeMarker>(100);
+    
+        public ShaderContextCache(Lifetime lifetime, ISolution solution, IPersistentIndexManager persistentIndexManager,
+            InjectedHlslFileLocationTracker locationTracker, DocumentManager manager, ILogger logger)
+        {
+            mySolution = solution;
+            myLocationTracker = locationTracker;
+            myManager = manager;
+            myLogger = logger;
+        }
+
+
+        public void SetContext(IPsiSourceFile psiSourceFile, CppFileLocation root)
+        {
+            using (ReadLockCookie.Create())
+            {
+                Assertion.Assert(root.RootRange.IsValid, "root.RootRange.IsValid()");
+                var range =  myManager.CreateRangeMarker(new DocumentRange(root.GetDocument(mySolution), root.RootRange));
+                myShaderContext.AddToCache(psiSourceFile, range);
+            }
+
+            var solution = psiSourceFile.GetSolution();
+            var psiFiles = solution.GetPsiServices().Files;
+
+            psiFiles.InvalidatePsiFilesCache(psiSourceFile);
+            solution.GetComponent<IDaemon>().Invalidate();
+        }
+
+        public CppFileLocation GetPreferredRootFile(CppFileLocation currentFile)
+        {
+            using (ReadLockCookie.Create())
+            {
+                var sourceFile = currentFile.GetRandomSourceFile(mySolution);
```

Comment:
I don't know how to fix it properly, it could happen only for player projects. I could use file system path for key, but after refactoring context will be lost. What do you think?

Reply:
What about adding it into the cache for all related source files? It means multiple entries, but it would be the same value. I don't know. It feels like an edge case, but I don't know how likely it would be, or what the impact is.

Reply:
I have decided to use filesystempath for now